### PR TITLE
[exporter/mezmoexporter] Add the Mezmo Log exporter

### PR DESCRIPTION
**Pull request by billmeyer**
**Description:** 

This adds a new exporter capable of exporting logs to the LogDNA/Mezmo platform as outlined in [9599](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9599).

Only the `log` telemetry type is supported at this time.

**Link to tracking Issue:** [9599](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9599)

**Testing:**

Unit test cases have been developed to test the bulk of the available interactions.  Testing coverage is presently at 93.4%:

```
opentelemetry-collector-contrib/exporter/mezmoexporter$ go test ./... -cover
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/mezmoexporter	0.567s	coverage: 93.4% of statements
```

**Documentation:**

`README.md` file is available with sample configuration entries needed to use the exporter with LogDNA.

### CODE REVIEW

File: exporter/mezmoexporter/README.md
Comment (dashpole) [inline exporter/mezmoexporter/README.md:8]: Should this link to to https://docs.logdna.com/docs/ingestion-key?

File: exporter/mezmoexporter/README.md
Comment (dashpole) [inline exporter/mezmoexporter/README.md:7]: Looking at the examples and default values, I don't think the endpoint is the same as the region?

File: exporter/mezmoexporter/config.go
Comment (dashpole) [inline exporter/mezmoexporter/config.go:55]: This doesn't match the docs above.  Lets be consistent on either `endpoint` or `ingest_url`

File: exporter/mezmoexporter/config.go
Comment (dashpole) [inline exporter/mezmoexporter/config.go:62]: I don't think this function needs to be exported.

File: exporter/mezmoexporter/config.go
Comment (dashpole) [inline exporter/mezmoexporter/config.go:84]: Is this function needed?

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [inline exporter/mezmoexporter/exporter.go:38]: This struct and its fields don't need to be exported.

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [inline exporter/mezmoexporter/exporter.go:46]: This struct and its fields don't need to be exported.

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [inline exporter/mezmoexporter/exporter.go:59]: I don't think this `wg` is neccessary.  stop() just needs to terminate background goroutines.

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [inline exporter/mezmoexporter/exporter.go:73]:
```suggestion
	m.client, err := m.config.HTTPClientSettings.ToClient(host.GetExtensions(), m.settings)
	return err
```

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [inline exporter/mezmoexporter/exporter.go:90]: remove commented out line

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [inline exporter/mezmoexporter/exporter.go:95]:
```suggestion
				log := logs.At(k)
```
nit: use conventional golang patterns

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [inline exporter/mezmoexporter/exporter.go:98]:
```suggestion
				attrs := map[string]string{}
```
nit: use conventional golang patterns

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [inline exporter/mezmoexporter/exporter.go:107]:
```suggestion
				s, _ := log.Attributes().Get("appname")
```
nit: use conventional golang patterns

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [inline exporter/mezmoexporter/exporter.go:108]:
```suggestion
				app := s.StringVal()
```
nit: use conventional golang patterns

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [inline exporter/mezmoexporter/exporter.go:110]:
```suggestion
				line := MezmoLogLine{
```
nit: use conventional golang patterns

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [inline exporter/mezmoexporter/exporter.go:124]: I don't love writing out json string parts here.  Is there a way to use a json marshaller to accomplish this?

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [inline exporter/mezmoexporter/exporter.go:159]: what is hostname meant to do?  "otel" seems like an odd value for it.  Should we be setting user-agent or something else instead?

File: exporter/mezmoexporter/utils.go
Comment (dashpole) [inline exporter/mezmoexporter/utils.go:31]: can we move this to a `_test.go` file so its clear it is used for testing only?

**Review by dashpole — COMMENTED** [review]
Please add an entry to CODEOWNERS with myself, and the two logdna owners as we discussed.

File: exporter/mezmoexporter/README.md
Comment (billmeyer) [reply exporter/mezmoexporter/README.md:3]: We can.  This will be changing very soon but for now I will change it to LogDNA.

File: exporter/mezmoexporter/README.md
Comment (billmeyer) [reply exporter/mezmoexporter/README.md:8]: Yes, corrected.

File: exporter/mezmoexporter/README.md
Comment (billmeyer) [reply exporter/mezmoexporter/README.md:7]: `endpoint` should have been `ingest_url`.   Corrected.

File: exporter/mezmoexporter/config.go
Comment (billmeyer) [reply exporter/mezmoexporter/config.go:55]: You are correct, they should all be `ingest_url` consistently.  Corrected.

File: exporter/mezmoexporter/config.go
Comment (billmeyer) [reply exporter/mezmoexporter/config.go:84]: No, it has been removed.

File: exporter/mezmoexporter/exporter.go
Comment (billmeyer) [reply exporter/mezmoexporter/exporter.go:38]: Corrected.

File: exporter/mezmoexporter/exporter.go
Comment (billmeyer) [reply exporter/mezmoexporter/exporter.go:46]: Corrected.

File: exporter/mezmoexporter/exporter.go
Comment (billmeyer) [reply exporter/mezmoexporter/exporter.go:84]: Removed.

File: exporter/mezmoexporter/utils.go
Comment (billmeyer) [reply exporter/mezmoexporter/utils.go:31]: Good catch.  Will move to `utils_test.go`.

File: exporter/mezmoexporter/exporter.go
Comment (billmeyer) [reply exporter/mezmoexporter/exporter.go:153]: This is a temporary fix as `hostname` is a required field in the URI of the POST.  Once we remove this requirement from the backend, we will remove this as well.  The hostname attributes attached to each OTEL log entry are being passed along as metadata to our platform and are being utilized appropriately as the source that tagged the data intended.

File: exporter/mezmoexporter/exporter.go
Comment (billmeyer) [reply exporter/mezmoexporter/exporter.go:118]: I'm not a fan either but given the way we are iterating over the variable length log entries and chunking them up into our JSON structure, I considered this to be the lesser of two evils.  It's a trade off of speed and efficiency vs overhead of using the marshaller and the compexity it creates in the code. The purpose is to accumulate the maximum payload size and when it exceeds 10MB, send what's accumulated and then proceed with the remaining log entries that are available.

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [reply exporter/mezmoexporter/exporter.go:152]: I'd suggest adding a TODO explaining why it is needed and when it can be removed.

If you wanted to, you could use the `host.name` resource attribute to fill this out.  But that's up to you.

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [reply exporter/mezmoexporter/exporter.go:117]: Got it, thanks.

File: exporter/mezmoexporter/config.go
Comment (dashpole) [inline exporter/mezmoexporter/config.go:65]: does this need to be exported?

Comment (billmeyer) [reply exporter/mezmoexporter/config.go:65]: Yes or the factory method createDefaultConfig() will fail because the return type is a`config.Exporter` interface which requires an exported `Validate()` function.

See interface definition in https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/exporter.go#L61.

File: exporter/mezmoexporter/exporter.go
Comment (billmeyer) [reply exporter/mezmoexporter/exporter.go:152]: Added.

File: exporter/mezmoexporter/config.go
Comment (dashpole) [reply exporter/mezmoexporter/config.go:65]: right.

File: exporter/mezmoexporter/README.md
Comment (codeboten) [inline exporter/mezmoexporter/README.md:22]: These values should be assigned with `:` instead of `=`

File: exporter/mezmoexporter/README.md
Comment (codeboten) [inline exporter/mezmoexporter/README.md:7]: If there's a default, why have it as a required field?

**Review by codeboten — COMMENTED** [review]
Just a few questions, please resolve the conflicts

File: exporter/mezmoexporter/doc.go
Comment (billmeyer) [reply exporter/mezmoexporter/doc.go]: Done.

File: exporter/mezmoexporter/README.md
Comment (billmeyer) [reply exporter/mezmoexporter/README.md]: Modified to allow this value as optional with the default URL supplied if omitted.

File: exporter/mezmoexporter/README.md
Comment (billmeyer) [reply exporter/mezmoexporter/README.md]: Fixed.

File: .github/CODEOWNERS
Comment (dashpole) [inline .github/CODEOWNERS:47]: Looks like this line was added during the rebase?

File: go.mod
Comment (dashpole) [inline go.mod:612]: Here as well

File: .github/CODEOWNERS
Comment (billmeyer) [reply .github/CODEOWNERS:47]: Resolved

Comment (dashpole) [reply .github/CODEOWNERS:47]: This line still needs to be removed.

Comment (billmeyer) [reply .github/CODEOWNERS]: Not sure why that was in there again but I've removed it and pushed the change.

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [reply exporter/mezmoexporter/exporter.go:59]: Still needs to be resolved.

File: exporter/mezmoexporter/exporter.go
Comment (dashpole) [inline exporter/mezmoexporter/exporter.go:92]: Do you still want to set this if the trace ID is empty?

Comment (dashpole) [reply exporter/mezmoexporter/exporter.go:92]: Same for span ID below

Comment (billmeyer) [reply exporter/mezmoexporter/exporter.go:92]: It's fine for now but we'll add the check in a future commit.  We have a couple other changes to add after this PR is merged.

**Review by dashpole — APPROVED** [review]

File: exporter/mezmoexporter/exporter_test.go
Comment (bogdandrutu) [inline exporter/mezmoexporter/exporter_test.go:30]: Use `go.opentelemetry.io/collector/semconv`

Comment (billmeyer) [reply exporter/mezmoexporter/exporter_test.go]: Corrected in [91b596c](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9743/commits/91b596c92311ca12c487673070232591a198ca9d).

File: CHANGELOG.md
Comment (codeboten) [inline CHANGELOG.md:103]:
```suggestion
- `mezmoexporter`: Add implementation of Mezmo Log exporter (#9743)
```

File: CHANGELOG.md
Comment (codeboten) [inline CHANGELOG.md:104]: This seems like it shouldn't be added here.

Comment (billmeyer) [reply CHANGELOG.md]: Fixed

File: CHANGELOG.md
Comment (billmeyer) [reply CHANGELOG.md]: Updated.

**Pull request merged**
